Clean/Table/Basic.lean

```diff
@@ -125,249 +125,249 @@ end TraceOfLength
-      assignment := fun x => if h : x >= ctx.offset && x < ctx.offset + size S then ⟨off, ⟨x-ctx.offset, by
-        simp only [ge_iff_le, Bool.and_eq_true, decide_eq_true_eq] at h
-        omega
-      ⟩⟩ else ctx.assignment x,
-      operations := ctx.operations ++ [GetRow off]
-    }
-
-  /-
-    Allocation of a sub-circuit moves the context offset by the witness length of the sub-circuit
-  -/
-  | Allocate subcircuit => {
-      offset := ctx.offset + subcircuit.local_length,
-      assignment := ctx.assignment,
-      operations := ctx.operations ++ [Allocate subcircuit]
-    }
+namespace TableContext
+@[reducible, table_norm, table_assignment_norm]
+def empty : TableContext W S F where
+  circuit := .from_offset 0
+  assignment := .empty W
 
-  /-
-    Assigning a variable to a cell in the trace just updates the assignment mapping
-  -/
-  | Assign v offset => {
-      offset := ctx.offset,
-      assignment := fun x => if x = v.index then offset else ctx.assignment x,
-      operations := ctx.operations ++ [Assign v offset]
-    }
+@[reducible, table_norm, table_assignment_norm]
+def offset (table : TableContext W S F) : ℕ := table.circuit.offset
+end TableContext
 
-instance {W: ℕ+} [Repr F] :
-    ToString (TableConstraintOperation W S F) where
-  toString
-    | Witness offset _ => "(Witness " ++ reprStr offset ++ ")"
-    | GetRow off => "(GetRow " ++ reprStr off ++ ")"
-    | Allocate {ops, ..} => "(Allocate " ++ reprStr ops ++ ")"
-    | Assign v offset => "(Assign " ++ reprStr v ++ " " ++ reprStr offset ++ ")"
+@[reducible, table_norm, table_assignment_norm]
+def TableConstraint (W: ℕ+) (S : Type → Type) (F : Type) [Field F] [ProvableType S] :=
+  StateM (TableContext W S F)
 
-end TableConstraintOperation
+@[table_assignment_norm]
+def assignment_from_circuit {n} (as: CellAssignment W S) : Operations F n → CellAssignment W S
+  | .empty _ => as
+  | .witness ops m _ => (assignment_from_circuit as ops).push_vars_aux m
+  | .assert ops _ => assignment_from_circuit as ops
+  | .lookup ops _ => assignment_from_circuit as ops
+  | .subcircuit ops s => (assignment_from_circuit as ops).push_vars_aux s.local_length
 
-@[reducible, table_norm]
-def TableConstraint (W: ℕ+) (S : Type → Type) (F : Type) [Field F] [NonEmptyProvableType S] :=
-  StateM (TableContext W S F)
+/--
+A `MonadLift` instance from `Circuit` to `TableConstraint` means that we can just use
+all circuit operations inside a table constraint.
+-/
+@[reducible, table_norm, table_assignment_norm]
+instance : MonadLift (Circuit F) (TableConstraint W S F) where
+  monadLift circuit ctx :=
+    let (a, ops) := circuit ctx.circuit
+    (a, {
+      circuit := ops,
+      -- the updated assignment is computed from a fresh starting circuit, independent of the circuit so far.
+      -- (if we would use `ops` instead of `circuit.operations 0`, we would be redoing previous assignments)
+      assignment := assignment_from_circuit ctx.assignment (circuit.operations 0)
```

Comment:
this is the main change! we can now use circuit operations directly inside a table